### DISCUSSION TITLE
Add type stubs to PyMuPDF -- IMPLEMENTED

### DISCUSSION THREAD

**Discussion by stangier** [pymupdf/PyMuPDF#786] [Ideas]
Hello!

I am currently using PyMuPDF in VSCode in conjunction with Pylance and strict type checking. Pylance reports missing type stubs when importing fitz. Have you considered adding type stubs to the PyMuPDF library? As this is an active developed library, I think this would be a great addition that would adapt it to newer Python conventions.

**Comment by JorjMcKie** [comment]
I looked into this just yesterday:
PyMuPDF does no longer support Python versions older than 3.6, so the path to something like that is free actually.
As you probably will know, the interface code to the C base library MuPDF is generated by SWIG.
SWIG can be told to generate type annotations automatically for functions written in C. That's good.
The possible problem is that it has limited knowledge about parameter and return code meanings and simply converts the C types to annotations in Python - like that:

```python
    def update_file(
        self,
        buffer: "PyObject *" = None,  # SWIG cannot know that this should be a bytes object
        filename: "char *" = None,  # should actually be string
        ufilename: "char *" = None,
        desc: "char *" = None,
    ) -> "PyObject *":  # SWIG won't know that the return is always None
        """Update attached file."""
        CheckParent(self)

        return _fitz.Annot_update_file(self, buffer, filename, ufilename, desc)
```

For methods written in Python there is no problem of course - other than a lot of work!
